frontend/src/components/auth/PrivateRoute.tsx

```diff
@@ -1,33 +1,56 @@
-import React, { useContext } from "react";
+import React, { type ComponentType, useContext } from "react";
 import { Redirect, Route } from "react-router-dom";
 
 import { HOME_PAGE } from "../../constants/Routes";
 import AuthContext from "../../contexts/AuthContext";
 import type { Role } from "../../types/AuthTypes";
 import NotFound from "../pages/NotFound";
 
-type PrivateRouteProps = {
-  component: React.FC;
+import usePageTitle from "./usePageTitle";
+
+type PageProps = {
+  component: ComponentType;
+  roles: Role[];
+  title?: string;
+};
+
+type PrivateRouteProps = PageProps & {
   path: string;
   exact?: boolean;
-  roles: Role[];
 };
 
-const PrivateRoute: React.FC<PrivateRouteProps> = ({
-  exact = false,
-  path,
-  component,
-  roles,
-}: PrivateRouteProps) => {
+const Page = ({ component: RenderComponent, roles, title }: PageProps) => {
   const { authenticatedUser } = useContext(AuthContext);
 
-  if (!authenticatedUser) {
+  const isSignedIn = authenticatedUser !== null;
+  const isAuthorized = isSignedIn && roles.includes(authenticatedUser.role);
+
+  usePageTitle(
+    isAuthorized ? title : "Not Found",
+    !isSignedIn || typeof title === "undefined",
```

Comment:
why do we need both `typeof title === "undefined"` and `if (!title) return;` in `usePageTitle`? i think this check prevents "Not Found" from displaying (e.g. if teacher visits /admin).

Reply:
oops, my bad. I must have missed the existing check when I was refactoring :')